contracts/OlasMech.sol

```diff
@@ -268,27 +217,56 @@ abstract contract OlasMech is Mech, IErrorsMech, ImmutableStorage {
 
     /// @dev Delivers a request by a marketplace.
     /// @notice This function ultimately calls mech marketplace contract to finalize the delivery.
-    /// @param requestId Request id.
-    /// @param data Self-descriptive opaque data-blob.
+    /// @param requestIds Set of request ids.
+    /// @param datas Corresponding set of self-descriptive opaque delivery data-blobs.
     function deliverToMarketplace(
-        uint256 requestId,
-        bytes memory data
+        uint256[] memory requestIds,
+        bytes[] memory datas
     ) external onlyOperator {
         // Reentrancy guard
-        if (_locked > 1) {
+        if (locked) {
             revert ReentrancyGuard();
         }
-        _locked = 2;
+        locked = true;
 
-        // Request delivery
-        bytes memory requestData = _deliver(requestId, data);
+        // Check array sizes
+        if (requestIds.length == 0 || requestIds.length != datas.length) {
+            revert WrongArrayLength(requestIds.length, datas.length);
+        }
+
+        // Preliminary delivery processing
+        bytes[] memory deliveryDatas = _prepareDeliveries(requestIds, datas);
+
+        // Get finalized delivery rates
+        uint256[] memory deliveryRates = getFinalizedDeliveryRates(requestIds);
+
+        // Mech marketplace delivery finalization
+        // Some of deliveries might be front-run by other mechs, and thus only actually delivered ones are recorded
+        bool[] memory deliveredRequests = IMechMarketplace(mechMarketplace).deliverMarketplace(requestIds,
```

Comment:
This is key - get exactly requests that were delivered via the marketplace, account for them, and ignore all undelivered ones.